File: EN/aigateway/cap-15/block-13.lua
# Extracted from: LibroAIGateway/cap-15-rate-limiting.md
-- Atomic INCR with EXPIRE only on creation
local key = KEYS[1]
local ttl = tonumber(ARGV[1])
local current = redis.call('INCR', key)
if current == 1 then
    redis.call('EXPIRE', key, ttl)
end
return current

File: EN/aigateway/cap-17/block-13.py
# Extracted from: LibroAIGateway/cap-17-sso-scim-mfa.md
# gateway/app/services/mfa_service.py — verification with atomic rate limit
@classmethod
async def verify_code(cls, redis, user_id: int, code: str) -> dict:
    key = f"mfa:{user_id}"
    if not await redis.exists(key):
        return {"valid": False, "reason": "expired"}

    # INCR-FIRST: we increment before comparing (anti race condition)
    new_attempts = await redis.hincrby(key, "attempts", 1)
    if int(new_attempts) > 5:
        await redis.delete(key)
        return {"valid": False, "reason": "max_attempts"}

    stored_code = await redis.hget(key, "code") or ""
    if not secrets.compare_digest(code.strip(), stored_code):
        return {"valid": False, "reason": "invalid_code"}

    await redis.delete(key)  # single-use
    return {"valid": True, "reason": "ok"}
